### DISCUSSION TITLE
Add "change" event in embed mode with specific changes made to the diagram.

### DISCUSSION THREAD

**Discussion by tyagirajat200** [jgraph/drawio#4321] [Q&A] [answered]
* [x] I agree to follow the [Code of Conduct](https://github.com/jgraph/drawio/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for a feature request that matches the one I want to file, without success.

**Is your feature request related to a problem? Please describe.**
Yes, Explained below.

**Describe the solution you'd like**
I am seeking a solution to implement real-time collaboration using Firebase Realtime Database with Draw.io in embedded mode. Currently, I am able to capture the "autosave" event which provides the entire XML of the current diagram view when changes occur. However, I aim to optimize this by capturing and transmitting only the specific changes made to the diagram, rather than the entire XML, to Firebase. This would involve detecting and transmitting the delta changes (i.e., the difference between the previous state and the updated state) to ensure efficient synchronization among users. Upon receiving these changes at other users' ends, a "merge" event would be triggered to integrate the new changes seamlessly into the existing context.

**Describe alternatives you've considered**
The current approach involves saving and transmitting the complete XML each time a change occurs, followed by a full reload of the diagram at the receiving end using the updated XML. However, this method disrupts the user experience by resetting the layout and focus of the diagram, causing the user's view to jump to the beginning of the diagram. Therefore, I am exploring alternative methods that enable capturing and transmitting only the incremental changes (delta) to minimize disruption and maintain user focus and layout continuity during collaboration sessions.

**Additional context**
NA

**Comment by LSViana** [comment]
Hi,

Is any event like this available in the embed mode of draw.io?

**Reply by LSViana** [reply]
Is there any maintainer that can provide some directions here?

I don't have enough context in the codebase but I'd be happy to provide a PR with this new feature.

**Comment by alderg** [comment]
Duplicate of https://github.com/jgraph/drawio/issues/2491

**Reply by LSViana** [reply]
Hi, @alderg!

Since #2491 is closed, what does it mean for this issue? Is it a valid use case to be implemented?

For example, if I'm using the draw.io embed and I need to know when the user changes the diagram, which of the events documented [here](https://www.drawio.com/doc/faq/embed-mode) should I use?

I noticed the `autosave` event is the closest one but it seems to be triggered when I push the `merge` action into the embed (to apply remote changes) as well.

**Comment by alderg** [comment, marked as answer]
Delta-based sync has since been added to the JSON embed protocol via the `diffSync` option of the `load` message (added in 29.6.5):

```js
iframe.contentWindow.postMessage(JSON.stringify(
	{action: 'load', xml: xml, autosave: 1, diffSync: true}), '*');
```

With this enabled:

- `autosave` events contain a `patch` with the changes since the last event and a `checksum` of the resulting state. The full XML is still included unless you use `diffSync: {patchOnly: true}`. Changes that produce no page diff (eg. page format) fall back to sending `xml` only, so handle both cases.
- The `patch` action applies a remote delta in-place, without reloading the editor: `{action: 'patch', patch: patch, checksum: checksum}`. The editor responds with `{event: 'patch', checksum: ...}` and adds `checksumMismatch: true` if the local result doesn't match the given checksum.
- `resetDiff` re-baselines the sync state, optionally merging the given `xml` first — use this to resync after a checksum mismatch. `getDiff` returns the pending changes as a patch on request, if you prefer polling over autosave.

Patches use draw.io's internal diff format and should be treated as opaque, ie. stored and relayed as-is (via Firebase in your case).

**Note on versions:** `diffSync` requires 29.6.5 or later. For concurrent editing you should use **30.2.7 or later**: since 30.2.7 each embedded editor assigns a unique ID prefix to new cells, so inserts from different clients don't collide. In earlier versions all embedded editors number new cells `2, 3, ...`, so the diff treats different cells created in different clients as the same cell and merges break. embed.diagrams.net is always on the latest version; this only matters if you self-host.

Minimal example with two editors kept in sync via patches — replace the direct relay with your Firebase pub/sub:

```html
<!DOCTYPE html>
<html>
<body>
<iframe id="a" src="https://embed.diagrams.net/?embed=1&proto=json&spin=1" width="45%" height="480"></iframe>
<iframe id="b" src="https://embed.diagrams.net/?embed=1&proto=json&spin=1" width="45%" height="480"></iframe>
<script>
var xml = '<mxGraphModel><root><mxCell id="0"/><mxCell id="1" parent="0"/></root></mxGraphModel>';
var a = document.getElementById('a').contentWindow;
var b = document.getElementById('b').contentWindow;

window.addEventListener('message', function(evt)
{
	if (evt.source != a && evt.source != b)
	{
		return;
	}

	var msg = JSON.parse(evt.data);
	var other = (evt.source == a) ? b : a;

	if (msg.event == 'init')
	{
		evt.source.postMessage(JSON.stringify({action: 'load',
			xml: xml, autosave: 1, diffSync: {patchOnly: true}}), '*');
	}
	else if (msg.event == 'autosave')
	{
		if (msg.patch != null)
		{
			// In a real app: publish the patch to Firebase here
			other.postMessage(JSON.stringify({action: 'patch',
				patch: msg.patch, checksum: msg.checksum}), '*');
		}
		else if (msg.xml != null)
		{
			// Change without a page diff: sync the full XML
			other.postMessage(JSON.stringify({action: 'merge',
				xml: msg.xml}), '*');
		}
	}
	else if (msg.event == 'patch' && msg.checksumMismatch)
	{
		// Clients have diverged: resync from your authoritative copy
		// with {action: 'resetDiff', xml: lastKnownGoodXml}
		console.log('Checksum mismatch, resync required');
	}
});
</script>
</body>
</html>
```
